From 3c2b7de1b0031ea304c670b28c3c7d9b65982943 Mon Sep 17 00:00:00 2001
From: miaohong chen <miaohong.chen@amlogic.com>
Date: Fri, 27 Aug 2021 19:27:49 +0800
Subject: [PATCH] v4l2: CB1 vp9 mosiac when seeking. [1/1]

PD#SWPL-57739

Problem:
When decoding the current vp9 frame,
it depends on the previous frame and the comv of the previous frame.
When abnormal, the frame buffer of the previous frame is released,
resulting in a blurry screen.

Solution:
Ensure that two adjacent frames do not use the same frame buffer.

Change-Id: I70fba53e08eabb28233383b2b07f260f2b45006b
Signed-off-by: miaohong chen <miaohong.chen@amlogic.com>
---
 drivers/frame_provider/decoder_v4l/vp9/vvp9.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/frame_provider/decoder_v4l/vp9/vvp9.c b/drivers/frame_provider/decoder_v4l/vp9/vvp9.c
index 3ad4668d..6986f870 100644
--- a/drivers/frame_provider/decoder_v4l/vp9/vvp9.c
+++ b/drivers/frame_provider/decoder_v4l/vp9/vvp9.c
@@ -2422,7 +2422,8 @@ static int get_free_fb_idx(struct VP9Decoder_s *pbi)
 
 	for (i = 0; i < pbi->used_buf_num; ++i) {
 		if ((frame_bufs[i].ref_count == 0) &&
-			(frame_bufs[i].buf.vf_ref == 0))
+			(frame_bufs[i].buf.vf_ref == 0) &&
+			(cm->cur_frame != &frame_bufs[i]))
 			break;
 	}
 
@@ -2452,7 +2453,8 @@ static int v4l_get_free_fb(struct VP9Decoder_s *pbi)
 			if ((frame_bufs[i].ref_count == 0) &&
 				(pic->vf_ref == 0) &&
 				(pic->index != -1) &&
-				pic->cma_alloc_addr) {
+				pic->cma_alloc_addr &&
+				(cm->cur_frame != &frame_bufs[i])) {
 				free_pic = pic;
 			}
 			break;
-- 
2.29.2

